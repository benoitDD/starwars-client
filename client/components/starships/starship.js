import React, {Component} from 'react'
import {Query} from 'react-apollo'
import {Starship as StarshipQuery} from '../../query/starships.gql'
import PropTypes from 'prop-types'
import Properties from '../utils/properties'
import {Pilots} from '../utils/linkObjects'
import Images from '../utils/images'
import AddImageHeader from '../utils/addImageHeader'
import '../objects.sass'
import Loading from '../utils/loading'
import HandleError from '../utils/handleError'

function getProperties(starship){
	return [
		{
			label: 'Model',
			value: starship.model
		},
		{
			label: 'Starship class',
			value: starship.starshipClass
		},
		{
			label: 'Manufacturers',
			value: starship.manufacturers
		},
		{
			label: 'Cost in credits',
			value: starship.costInCredits,
			unit: 'credits'
		},
		{
			label: 'Length',
			value: starship.length,
			unit: 'm'
		},
		{
			label: 'Crew',
			value: starship.crew,
			helper: 'The number of personnel needed to run or pilot this starship'
		},
		{
			label: 'Passengers',
			value: starship.passengers,
			helper: 'The number of non-essentiel people this starship can transport'
		},
		{
			label: 'Max atmosphering speed',
			value: starship.maxAtmospheringSpeed,
			helper: 'The maximum speed of this starship in atmosphere'
		},
		{
			label: 'Hyperdrive rating',
			value: starship.hyperdriveRating,
			helper: 'The class of this starships hyperdrive'
		},
		{
			label: 'MGLT',
			value: starship.MGLT,
			helper: 'The Maximum number of Megalights this starship can travel in a standard hour'
		},
		{
			label: 'cargo capacity',
			value: starship.cargoCapacity,
			unit: 'kg'
		},
		{
			label: 'Consumables',
			value: starship.consumables,
			helper: 'The maximum length of time that this starship can provide consumables for its entire crew without having to resupply'
		}
	]
}

class Starship extends Component {
	render(){
		return (
			<Query query = {StarshipQuery} variables = {{id:this.props.id}} fetchPolicy = "network-only">
				{
					({data, loading, error}) => {
						if (loading) return <div className = 'object_loading'><Loading/></div>
						if (error) return <HandleError error = {error}/>
						const {starship} = data
						const properties = getProperties(starship)
						return (
							<section>
								<h1>{starship.name}</h1>
								<article>
									<Images images = {starship.more && starship.more.imagesHeader} idOwn = {starship.id} />
									<AddImageHeader id = {starship.id} typeObject = 'Starship' />
									<div className = 'object_descriptions'>
										<Properties properties = {properties}/>
										<Pilots pilots = {starship.persons ? starship.persons.persons : undefined} />
									</div>
								</article>
							</section>
						)
					}
				}
			</Query>
		)
	}
}

Starship.propTypes = {
	id: PropTypes.string,
	path: PropTypes.string
}

export default Starship