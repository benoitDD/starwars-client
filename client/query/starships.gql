#import './fragments.gql'

query AllStarships($pageSize: Int, $after: String, $before: String){
  allStarships(pageSize: $pageSize, after: $after, before: $before) {
    pageInfo {
      endCursor
      startCursor
      hasNextPage
      hasPreviousPage
      totalCount
    }
    starships {
      name
      id
    }
  }
}

query Starship($id: ID!){
  starship(id: $id) {
    ...StarshipFragmentMore
  }
}

fragment StarshipFragment on Starship {
  id
  name
  model
  starshipClass
  manufacturers
  costInCredits
  length
  crew
  passengers
  maxAtmospheringSpeed
  hyperdriveRating
  MGLT
  cargoCapacity
  consumables
}