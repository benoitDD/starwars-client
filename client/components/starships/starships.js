import React, {Component} from 'react'
import {LinkStarship} from '../utils/links'
import {AllStarships} from '../../query/starships.gql'
import Pagination from '../utils/pagination'

class Starships extends Component {
	render(){
		return (
			<Pagination query = {AllStarships} nameQuery = 'allStarships' nameObjects = 'starships'>
				{
					starship => {
						return (
							<LinkStarship id = {starship.id}>
								{starship.name}
							</LinkStarship>
						)
					}
				}
			</Pagination>
		)
	}
}

export default Starships