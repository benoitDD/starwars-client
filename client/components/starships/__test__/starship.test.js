import React from 'react'
import renderer from 'react-test-renderer'
import {MockedProvider} from 'react-apollo/test-utils'
import wait from 'waait'
import {Starship} from '../../../query/starships.gql'
import StarshipComponent from '../starship'
import dataStarship from '../../test/starship.json'

jest.mock('../../../hoc/updateActive.js')
jest.mock('../../../hoc/componentPrivate.js')
jest.mock('../../../context.js')

const id = dataStarship.data.starship.id

const mocks = [
	{
		request: {
			query: Starship,
			variables: {
				id
			}
		},
		result: dataStarship
	}
]

test('Test Starship', async (done) => {
	const component = renderer.create(
		<MockedProvider mocks={mocks}>
			<StarshipComponent id = {id}/>
		</MockedProvider>,
	)

	await wait(0)
    
	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()
	done()
})

test('Test Starship loading',async (done) => {
	const component = renderer.create(
		<MockedProvider mocks={mocks}>
			<StarshipComponent id = {id}/>
		</MockedProvider>,
	)
    
	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()
	done()
})

test('Test Starship error', async (done) => {
	const mocks = [
		{
			request: {
				query: Starship,
				variables: {
					id
				}
			},
			result: dataStarship,
			error: new Error('Une erreur :-)')
		}
	]

	const component = renderer.create(
		<MockedProvider mocks={mocks}>
			<StarshipComponent id = {id}/>
		</MockedProvider>,
	)

	await wait(0)
    
	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()
	done()
})