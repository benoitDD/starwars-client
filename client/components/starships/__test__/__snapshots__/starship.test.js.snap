// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Starship 1`] = `
<section>
  <h1>
    X-wing
  </h1>
  <article>
    <div
      className="images"
    >
      <div
        className="images-item"
      >
        <span
          className="remove-image"
          onClick={[Function]}
        >
          ‚ùå
        </span>
        <img
          alt="X-wing"
          className="images-item-image"
          src="undefined/images/eae03a93.jpeg"
          title="X-wing"
        />
      </div>
    </div>
    <form
      className="add-image-form"
    >
      <div
        className="add-image-form-items"
      >
        <div
          className="add-image-form-item"
        >
          <div
            className="add-image-form-item-inputs"
          >
            <label
              htmlFor="title"
            >
              title
            </label>
            <input
              id="title"
              name="title"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="add-image-form-error"
          />
        </div>
        <div
          className="add-image-form-item"
        >
          <div
            className="add-image-form-item-inputs"
          >
            <label
              htmlFor="description"
            >
              description
            </label>
            <input
              id="description"
              name="description"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="add-image-form-error"
          />
        </div>
        <div
          className="add-image-form-item"
        >
          <div
            className="add-image-form-item-inputs"
          >
            <label
              htmlFor="image"
            >
              image
            </label>
            <input
              accept="image/png, image/jpeg, image/jpg"
              id="image"
              type="file"
            />
          </div>
          <div
            className="add-image-form-error"
          />
        </div>
        <button
          onClick={[Function]}
          type="submit"
        >
          add
        </button>
      </div>
      <div />
    </form>
    <div
      className="object_descriptions"
    >
      <div
        className="properties"
      >
        <h2
          className="properties-title"
        >
          properties
        </h2>
        <ul
          className="properties-attributes"
        >
          <li
            className="peer"
          >
            <div
              className="propertie-label"
            >
              model
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                T-65 X-wing
              </div>
              
            </div>
          </li>
          <li
            className="odd"
          >
            <div
              className="propertie-label"
            >
              starship.class
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                Starfighter
              </div>
              
            </div>
          </li>
          <li
            className="peer"
          >
            <div
              className="propertie-label"
            >
              manufacturers
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                <ul
                  className="propertie-value-array"
                >
                  <li
                    className="propertie-value-array-item peer"
                  >
                    Incom Corporation
                  </li>
                </ul>
              </div>
              
            </div>
          </li>
          <li
            className="odd"
          >
            <div
              className="propertie-label"
            >
              cost.in.credits
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                149999
              </div>
              <div
                className="properties-unit"
              >
                credits
              </div>
            </div>
          </li>
          <li
            className="peer"
          >
            <div
              className="propertie-label"
            >
              length
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                12.5
              </div>
              <div
                className="properties-unit"
              >
                m
              </div>
            </div>
          </li>
          <li
            className="odd helper"
            title="crew.helper"
          >
            <div
              className="propertie-label"
            >
              crew
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                1
              </div>
              
            </div>
          </li>
          <li
            className="peer helper"
            title="passengers.helper"
          >
            <div
              className="propertie-label"
            >
              passengers
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                0
              </div>
              
            </div>
          </li>
          <li
            className="odd helper"
            title="max.atmosphering.speed.helper"
          >
            <div
              className="propertie-label"
            >
              max.atmosphering.speed
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                1050
              </div>
              
            </div>
          </li>
          <li
            className="peer helper"
            title="hyperdrive.rating.helper"
          >
            <div
              className="propertie-label"
            >
              hyperdrive.rating
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                1
              </div>
              
            </div>
          </li>
          <li
            className="odd helper"
            title="mglt.helper"
          >
            <div
              className="propertie-label"
            >
              mglt
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                100
              </div>
              
            </div>
          </li>
          <li
            className="peer"
          >
            <div
              className="propertie-label"
            >
              cargo.capacity
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                110
              </div>
              <div
                className="properties-unit"
              >
                kg
              </div>
            </div>
          </li>
          <li
            className="odd helper"
            title="consumables.helper"
          >
            <div
              className="propertie-label"
            >
              consumables
            </div>
            <div
              className="propertie-value-unit"
            >
              <div
                className="propertie-value"
              >
                1 week
              </div>
              
            </div>
          </li>
        </ul>
      </div>
      <div
        className="link-object"
      >
        <h2
          className="link-object-title"
        >
          pilots
        </h2>
         
        <ul
          className="link-object-list"
        >
          <li>
            <a
              className="link"
              href="/person/cGVvcGxlOjE="
              onClick={[Function]}
            >
              <span
                className="link-object-item"
              >
                <span>
                  Luke Skywalker
                </span>
                <span>
                  <img
                    alt="luke skywalker"
                    src="undefined/images/78092401.jpeg"
                    title="luke skywalker"
                  />
                </span>
              </span>
            </a>
          </li>
          <li>
            <a
              className="link"
              href="/person/cGVvcGxlOjk="
              onClick={[Function]}
            >
              <span
                className="link-object-item"
              >
                <span>
                  Biggs Darklighter
                </span>
                <span>
                  <img
                    alt="Biggs Darklighter"
                    src="undefined/images/cc018650.jpeg"
                    title="Biggs Darklighter"
                  />
                </span>
              </span>
            </a>
          </li>
          <li>
            <a
              className="link"
              href="/person/cGVvcGxlOjE4"
              onClick={[Function]}
            >
              <span
                className="link-object-item"
              >
                <span>
                  Wedge Antilles
                </span>
                <span
                  className="link-object-item-not-image"
                >
                  Any images define
                </span>
              </span>
            </a>
          </li>
          <li>
            <a
              className="link"
              href="/person/cGVvcGxlOjE5"
              onClick={[Function]}
            >
              <span
                className="link-object-item"
              >
                <span>
                  Jek Tono Porkins
                </span>
                <span
                  className="link-object-item-not-image"
                >
                  Any images define
                </span>
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </article>
</section>
`;

exports[`Test Starship error 1`] = `
<div>
  Une erreur inconnue s'est produite 
  : Network error: Une erreur :-)
</div>
`;

exports[`Test Starship loading 1`] = `
<div
  className="object_loading"
>
  
</div>
`;
